File: 16-3sum-closest/16-3sum-closest.java
class Solution {
    public int threeSumClosest(int[] nums, int target){
        Arrays.sort(nums);
        
        int len = nums.length-2;
        int fullLen = len+1;
        int sum = Integer.MAX_VALUE;


        for(int i= 0, j = 1, k = fullLen; i < len; ++i, j = i+1, k = fullLen){
            int absSum = Math.abs(sum - target);
            while(j < k){
                int newSum = nums[i] + nums[j] + nums[k];
                if(Math.abs(newSum - target) < absSum){
                    sum = newSum;
                    absSum = Math.abs(sum - target);
                }
                if(target > newSum) ++j;
                else if(target < newSum) --k;
                else return newSum;
            }
        }

        return sum;
    }
}

File: 16-3sum-closest/16-3sum-closest.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        
        int closest = nums[0]+nums[1]+nums[2];
        sort(nums.begin(),nums.end());
        for(int i=0;i<nums.size()-2;i++)
        {
            if(i == 0 || nums[i] != nums[i-1])
            {
                int j = i+1;
                int k = nums.size()-1;
                
                while(j<k)
                {
                    int sum = nums[i]+nums[j]+nums[k];
                    if(sum == target)return sum;
                    if(abs(target - sum) < abs(target-closest))
                    {
                        closest = sum;
                    }
                    else if(target > sum )
                    {
                        j++;
                    }
                    else
                        k--;
                }
            }
        }
        return closest;
        
    }
};